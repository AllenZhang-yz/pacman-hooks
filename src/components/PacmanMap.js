import React from "react";
import Pacman from "./Pacman";

class PacmanMap extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      girdList: [],
      isShowPacman: false
    };
  }

  componentDidMount() {
    const { gridNumber } = this.props;
    let newGridList = [];
    newGridList.length = gridNumber * gridNumber;
    newGridList.fill("");
    this.setState({
      girdList: newGridList
    });
  }

  isPacmanPosition = index => {
    const { gridNumber, position} = this.props;
    const pacmanIndex = position.y * gridNumber + position.x;

    return index === pacmanIndex;
  };

  render() {
    const { direction, gridNumber } = this.props;
    const { girdList } = this.state;

    return (
      <div className="pacman-map">
        {girdList.map((grid, index) => (
          <div key={index} className="pacman-map__cell">
            {`(${index % gridNumber}, ${parseInt(index / gridNumber)})`}
            {this.isPacmanPosition(index) ? (
              <Pacman direction={direction} />
            ) : null}
          </div>
        ))}
      </div>
    );
  }
}

export default PacmanMap;
