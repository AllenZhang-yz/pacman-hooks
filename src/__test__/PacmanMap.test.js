import React from "react";
import PacmanMap from "../components/PacmanMap";
import { shallow } from "enzyme";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
Enzyme.configure({ adapter: new Adapter() });

const setup = ({ ...props }) => {
  const wrapper = shallow(<PacmanMap {...props} />);
  return {
    props,
    wrapper
  };
};

describe("PacmanMap", () => {
  const { props, wrapper } = setup({
    gridNumber: 5,
    direction: "NORTH",
    position: { x: 2, y: 3 }
  });
  it("PacmanMap renders correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it("isPacmanPosition correctly", () => {
    const testIndex = props.position.y * props.gridNumber + props.position.x;
    expect(wrapper.instance().isPacmanPosition(testIndex)).toBe(true);
    expect(wrapper.instance().isPacmanPosition(testIndex + 1)).toBe(false);
  })
  it(`PacmanMap's number is same to ${props.gridNumber} *${props.gridNumber}`, () => {
    const grids = wrapper.find('.pacman-map__cell');
    expect(grids.length).toBe(25);
  })
  it("PacmanMap description correctly", () => {
    const description = wrapper.find('p').text();
    expect(description).toBe(`Grid Number: ${props.gridNumber}`);
  })
});
