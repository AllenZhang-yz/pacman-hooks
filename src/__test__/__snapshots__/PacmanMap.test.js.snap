// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PacmanMap PacmanMap renders correctly 1`] = `
<div>
  <p>
    Grid Number: 
    5
  </p>
  <div
    className="pacman-map"
    style={
      Object {
        "width": "300px",
      }
    }
  >
    <div
      className="pacman-map__cell"
      key="0"
    >
      (0, 0)
    </div>
    <div
      className="pacman-map__cell"
      key="1"
    >
      (1, 0)
    </div>
    <div
      className="pacman-map__cell"
      key="2"
    >
      (2, 0)
    </div>
    <div
      className="pacman-map__cell"
      key="3"
    >
      (3, 0)
    </div>
    <div
      className="pacman-map__cell"
      key="4"
    >
      (4, 0)
    </div>
    <div
      className="pacman-map__cell"
      key="5"
    >
      (0, 1)
    </div>
    <div
      className="pacman-map__cell"
      key="6"
    >
      (1, 1)
    </div>
    <div
      className="pacman-map__cell"
      key="7"
    >
      (2, 1)
    </div>
    <div
      className="pacman-map__cell"
      key="8"
    >
      (3, 1)
    </div>
    <div
      className="pacman-map__cell"
      key="9"
    >
      (4, 1)
    </div>
    <div
      className="pacman-map__cell"
      key="10"
    >
      (0, 2)
    </div>
    <div
      className="pacman-map__cell"
      key="11"
    >
      (1, 2)
    </div>
    <div
      className="pacman-map__cell"
      key="12"
    >
      (2, 2)
    </div>
    <div
      className="pacman-map__cell"
      key="13"
    >
      (3, 2)
    </div>
    <div
      className="pacman-map__cell"
      key="14"
    >
      (4, 2)
    </div>
    <div
      className="pacman-map__cell"
      key="15"
    >
      (0, 3)
    </div>
    <div
      className="pacman-map__cell"
      key="16"
    >
      (1, 3)
    </div>
    <div
      className="pacman-map__cell"
      key="17"
    >
      (2, 3)
      <Pacman
        direction="NORTH"
      />
    </div>
    <div
      className="pacman-map__cell"
      key="18"
    >
      (3, 3)
    </div>
    <div
      className="pacman-map__cell"
      key="19"
    >
      (4, 3)
    </div>
    <div
      className="pacman-map__cell"
      key="20"
    >
      (0, 4)
    </div>
    <div
      className="pacman-map__cell"
      key="21"
    >
      (1, 4)
    </div>
    <div
      className="pacman-map__cell"
      key="22"
    >
      (2, 4)
    </div>
    <div
      className="pacman-map__cell"
      key="23"
    >
      (3, 4)
    </div>
    <div
      className="pacman-map__cell"
      key="24"
    >
      (4, 4)
    </div>
  </div>
</div>
`;
